README.md
# NewGamePlus

Allows you to start a new game plus once you launch a rocket. Generate a new world and leave your home world, but carry over an inventory full of items and your research.

Enabled the mod option to be able to run the ngp-gui command. Then simply type /ngp-gui in chat (or the console, it's the same thing), and you will see the gui and be able to start the new game plus without launching the rocket.

The above processes can be repeated an infinite amount of times. Please know that it is not possible to go back to your old world(s)! Pressing the start button might also freeze your game for up to a few minutes. That's fine. It'll unfreeze eventually.

Works with mods that add ores or terrain etc. If you want to generate a new world with them, first enable them, then start the new game plus, not the other way around.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 5.0.1
Date: 2022-01-27
  Features:
    - Added peaceful mode description.
    - Changed license to MIT.
      I cannot (reasonably) play games without controller support.
      Maintaining mods for a game I can't play is an uninteresting hobby for me.
      So I want to make it easier for others to fork my mods.
      May you take this code and have fun with it.
  Bugfixes:
    - Fixed locale of labels for vanilla ores.
    - Fixed that space exploration planet size was reset to default when starting a new game plus. 
---------------------------------------------------------------------------------------------------
Version: 5.0.0
Date: 2021-02-11
  Features:
    - Updated to Factorio 1.1.
---------------------------------------------------------------------------------------------------
Version: 4.0.2
Date: 26. 05. 2020
  Minor Features:
    - Updated for 0.18.27.
---------------------------------------------------------------------------------------------------
Version: 4.0.1
Date: 28. 04. 2020
  Minor Features:
    - Excluded some surfaces from Space Exploration and Mobile Factory from New Game+ reset.
---------------------------------------------------------------------------------------------------
Version: 4.0.0
Date: 25. 01. 2020
  Major Features:
    - Updated for 0.18.
---------------------------------------------------------------------------------------------------
Version: 3.1.2
Date: 13. 11. 2019
  Minor Features:
    - Added custom event so that other mods are notified after the new game plus is created.
---------------------------------------------------------------------------------------------------
Version: 3.1.1
Date: 13. 11. 2019
  Bugfixes:
    - Fixed that it was not detected correctly whether a new world had been generated.
---------------------------------------------------------------------------------------------------
Version: 3.1.0
Date: 16. 07. 2019
  Minor Features:
    - Used new functions to restrict text fields to numbers only.
    - Added a button to import max exchange strings.
    - Changed to reuse nauvis instead of creating its own surface, improves mod compatibility.
---------------------------------------------------------------------------------------------------
Version: 3.0.2
Date: 31. 05. 2019
  Bugfixes:
    - Fixed issue related to division by 0.
---------------------------------------------------------------------------------------------------
Version: 3.0.1
Date: 29. 05. 2019
  Minor Features:
    - Fixed RSO compatibility.
---------------------------------------------------------------------------------------------------
Version: 3.0.0
Date: 28. 05. 2019
  Major Features:
    - Updated for 0.17.
---------------------------------------------------------------------------------------------------
Version: 2.1.3
Date: 02. 12. 2018
  Minor Features:
    - Added custom event so that other mods are notified when I reset technologies.
---------------------------------------------------------------------------------------------------
Version: 2.1.2
Date: 14. 10. 2018
  Minor Features:
    - Added custom event so that other mods are notified when I delete nauvis.
---------------------------------------------------------------------------------------------------
Version: 2.1.1
Date: 16. 08. 2018
  Bugfixes:
    - Fixed that surface creation would fail under certain circumstances.
---------------------------------------------------------------------------------------------------
Version: 2.1.0
Date: 27. 01. 2018
  Minor Features:
    - Added cliff settings to the options.
  Bugfixes:
    - Fixed that the player didnt consider tile transitions.
---------------------------------------------------------------------------------------------------
Version: 2.0.1
Date: 03. 01. 2018
  Bugfixes:
    - Fixed that new players in multiplayers games would spawn in the wrong world.
---------------------------------------------------------------------------------------------------
Version: 2.0.0
Date: 11. 12. 2017
  Major Features:
    - Updated for 0.16.
---------------------------------------------------------------------------------------------------
Version: 1.1.2
Date: 04. 11. 2017
  Bugfixes:
    - Fixed building platform option appearing although the building platform mod was not installed.
    - Changed how on_entity_died is raised when deleting chunks, should prevent weirdness due to chunk borders when interacting with another mod's entities.
---------------------------------------------------------------------------------------------------
Version: 1.1.1
Date: 07. 10. 2017
  Minor Features:
    - Added support for the building platform mod.
    - Minor code cleanup.
---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 25. 09. 2017
  Major Features:
    - Added all advanced settings that can be found in the map generation GUI to the advanced settings.
    - Added option to reset enemy evolution.
    - Added "Reset to default" button that resets all settings back to the default.
    - The "Set to current map settings" button now also sets the advanced settings to the currently used settings.
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 18. 09. 2017
  Major Features:
    - Initial release.


gui.lua
local mod_gui = require("__core__/lualib/mod-gui")
local util = require("__NewGamePlus__/utilities")
local map_gen_gui = require("__NewGamePlus__/map_gen_settings_gui")
local map_settings_gui = require("__NewGamePlus__/map_settings_gui")

local gui = {}

gui.regen = function(player)
  gui.kill(player)
  --toggle button
  local button_flow = mod_gui.get_button_flow(player)
  local button = button_flow.add{
    type = "button",
    name = "new-game-plus-toggle-config",
    style = mod_gui.button_style,
    tooltip = {"gui.new-game-plus-toggle-tooltip", {"gui.new-game-plus-config-caption"}},
    caption = {"gui.new-game-plus-toggle-config-caption"}
  }
  button.visible = true
  -- general gui frame setup --
  local frame_flow = mod_gui.get_frame_flow(player)
  local config_frame = frame_flow.add{
    type = "frame",
    name = "new-game-plus-config-frame",
    direction = "vertical"
  }
  config_frame.visible = false
  -- it's the top flow of the main window, has title, import and advanced button
  local config_frame_title_flow = config_frame.add{
    type = "flow"
  }
  config_frame_title_flow.add{
    type = "label",
    caption = {"gui.new-game-plus-config-caption"},
    style = "frame_title"
  }
  config_frame_title_flow.add{
    type = "flow",
    style = "pusher"
  }
  -- advanced button
  local advanced_button = config_frame_title_flow.add{
    type = "button",
    name = "new-game-plus-more-options",
    style = mod_gui.button_style,
    tooltip = {"gui.new-game-plus-toggle-tooltip", {"gui-map-generator.advanced-tab-title"}},
    caption = {"gui-map-generator.advanced-tab-title"}
  }
  advanced_button.style.height = 24
  advanced_button.style.bottom_padding = 4

  --make main gui sections
  local inner_frame = config_frame.add{
    type = "frame",
    direction = "vertical",
    style = "deep_frame",
    name = "new-game-plus-config-inner-frame"
  }
  gui.make_basic_settings_gui(inner_frame)
  local config_subframe = inner_frame.add{
    type = "flow",
    name = "new-game-plus-config-subframe",
    direction = "horizontal"
  }
  map_gen_gui.create(config_subframe)

  -- advanced window
  local config_more_frame = frame_flow.add{
    type = "frame",
    caption = {"gui-map-generator.advanced-tab-title"},
    name = "new-game-plus-config-more-frame",
    direction = "vertical"
  }
  config_more_frame.visible = false
  gui.make_advanced_settings_gui(config_more_frame)

  -- start button at the bottom
  local start_button_flow = config_frame.add{
    type = "flow",
    direction = "horizontal"
  }
  start_button_flow.style.top_padding = 8
  start_button_flow.add{
    type = "flow",
    style = "pusher"
  }
  start_button_flow.add{
    type = "button",
    name = "new-game-plus-start-button",
    style = "confirm_button",
    tooltip = {"gui.new-game-plus-start-button-tooltip"},
    caption = {"", "[img=utility/warning]  ", {"gui.new-game-plus-start-button-caption"}}
  }
end

gui.kill = function(player)
  local button_flow = mod_gui.get_button_flow(player)
  local frame_flow = mod_gui.get_frame_flow(player)
  if button_flow["new-game-plus-toggle-config"] then
    button_flow["new-game-plus-toggle-config"].destroy()
  end
  if frame_flow["new-game-plus-config-frame"] then
    frame_flow["new-game-plus-config-frame"].destroy()
  end
  if frame_flow["new-game-plus-config-more-frame"] then
    frame_flow["new-game-plus-config-more-frame"].destroy()
  end
  gui.kill_mes_import_window(player)
end

gui.kill_mes_import_window = function(player)
  if player.gui.center["new-game-plus-import-frame"] then
    player.gui.center["new-game-plus-import-frame"].destroy()
  end
end

gui.make_basic_settings_gui = function(parent)
  do --subheader
    local frame = parent.add{
      type = "frame",
      direction = "horizontal",
      style = "subheader_frame"
    }
    frame.add{
      type = "flow",
      direction = "horizontal",
      style = "pusher"
    }
    -- map exchange string button
    frame.add{
      type = "sprite-button",
      name = "new-game-plus-map-exchange-string",
      style = "tool_button",
      tooltip = {"gui-map-generator.import-exchange-string-tt"},
      sprite = "utility/import_slot"
    }
    -- tool buttons
    frame.add{
      type = "sprite-button",
      name = "new-game-plus-use-current-button",
      style = "tool_button",
      sprite = "utility/refresh",
      tooltip = {"gui.new-game-plus-use-current-button-caption"}
    }
    frame.add{
      type = "sprite-button",
      name = "new-game-plus-default-button",
      style = "tool_button_red",
      sprite = "utility/reset",
      tooltip = {"gui.new-game-plus-default-button-caption"}
    }
  end

  local config_option_table = parent.add{
    type = "table",
    name = "new-game-plus-config-option-table",
    column_count = 5
  }
  config_option_table.style.left_padding = 8
  config_option_table.style.right_padding = 8
  config_option_table.add{
    type = "label",
    caption = {"gui.new-game-plus-reset-evo-caption"}
  }
  config_option_table.add{
    type = "checkbox",
    name = "new-game-plus-reset-evo-checkbox",
    state = true
  }
  config_option_table.add{
    type = "flow",
    style = "pusher"
  }
  config_option_table.add{
    type = "label",
    caption = {"gui.new-game-plus-seed-caption"}
  }
  config_option_table.add{
    type = "textfield",
    name = "new-game-plus-seed-textfield",
    text = "0",
    numeric = true,
    allow_decimal = false,
    allow_negative = false
  }
  config_option_table.add{
    type = "label",
    caption = {"gui.new-game-plus-reset-research-caption"}
  }
  config_option_table.add{
    type = "checkbox",
    name = "new-game-plus-reset-research-checkbox",
    state = false
  }
  config_option_table.add{
    type = "flow",
    style = "pusher"
  }
  config_option_table.add{
    type = "label",
    caption = {"gui.new-game-plus-width-caption"}
  }
  config_option_table.add{
    type = "textfield",
    name = "new-game-plus-width-textfield",
    text = "0",
    numeric = true,
    allow_decimal = false,
    allow_negative = false
  }
  config_option_table.add{
    type = "label",
    caption = util.add_info_icon_to_string({"gui-map-generator.peaceful-mode-checkbox"}),
    tooltip = {"gui-map-generator.peaceful-mode-description"}
  }
  config_option_table.add{
    type = "checkbox",
    name = "new-game-plus-peaceful-mode-checkbox",
    state = false
  }
  config_option_table.add{
    type = "flow",
    style = "pusher"
  }
  config_option_table.add{
    type = "label",
    caption = {"gui.new-game-plus-height-caption"}
  }
  config_option_table.add{
    type = "textfield",
    name = "new-game-plus-height-textfield",
    text = "0",
    numeric = true,
    allow_decimal = false,
    allow_negative = false
  }
end

gui.make_advanced_settings_gui = function(parent)
  local subframe = parent.add{
    type = "frame",
    direction = "vertical",
    style = "b_inner_frame",
    name = "new-game-plus-config-more-frame",
  }
  local table = subframe.add{
    type = "table",
    name = "new-game-plus-config-more-table",
    column_count = 2,
    vertical_centering = false
  }

  table.style.horizontal_spacing = 12
  local map_settings = game.map_settings
  --make different advanced option groups
  map_settings_gui.make_pollution_settings(table, map_settings)
  map_settings_gui.make_expansion_settings(table, map_settings)
  map_settings_gui.make_evolution_settings(table, map_settings)
  gui.make_difficulty_settings_gui(table)
end

gui.make_difficulty_settings_gui = function(parent)
  local config_more_option_difficulty_flow = parent.add{
    type = "flow",
    name = "new-game-plus-config-more-difficulty-flow",
    direction = "vertical"
  }
  config_more_option_difficulty_flow.add{
    type = "label",
    caption = {"gui.new-game-plus-recipes-and-technology-title"},
    style = "caption_label"
  }
  local table = config_more_option_difficulty_flow.add{
    type = "table",
    name = "new-game-plus-config-more-difficulty-table",
    column_count = 2,
    style = "bordered_table"
  }
  table.style.column_alignments[2] = "center"

  local difficulty_options = {{"recipe-difficulty.normal"}, {"recipe-difficulty.expensive"}}
  table.add{
    type = "label",
    caption = {"gui.new-game-plus-recipe-difficulty"}
  }
  local recipe_difficulty = table.add{
    type = "drop-down",
    name = "new-game-plus-recipe-difficulty-drop-down",
  }
  recipe_difficulty.items = difficulty_options
  recipe_difficulty.selected_index = game.difficulty_settings.recipe_difficulty + 1
  table.add{
    type = "label",
    caption = {"gui.new-game-plus-technology-difficulty"}
  }
  local technology_difficulty = table.add{
    type = "drop-down",
    name = "new-game-plus-technology-difficulty-drop-down",
  }
  technology_difficulty.items = difficulty_options
  technology_difficulty.selected_index = game.difficulty_settings.technology_difficulty + 1
  table.add{
    type = "label",
    caption = {"gui.new-game-plus-technology-price-multiplier"}
  }
  local technology_multiplier = table.add{
    type = "textfield",
    name = "new-game-plus-technology-multiplier-textfield",
    numeric = true,
    allow_decimal = true,
    allow_negative = false
  }
  technology_multiplier.text = tostring(game.difficulty_settings.technology_price_multiplier)
  technology_multiplier.style.maximal_width = 50
end

gui.open_mes_import_window = function(player)
  gui.kill_mes_import_window(player)
  local frame = player.gui.center.add{
    type = "frame",
    name = "new-game-plus-import-frame",
    direction = "vertical",
    caption = {"gui.map-exchange-string"}
  }
  frame.add{
    type = "label",
    caption = {"gui-map-generator.exchange-string-instructions"}
  }
  local textbox = frame.add{
    type = "text-box",
    name = "new-game-plus-import-text-box",
  }
  textbox.style.width = 500
  textbox.style.height = 250
  local button_flow = frame.add{
    type = "flow"
  }
  button_flow.add{
    type = "flow",
    style = "pusher"
  }
  button_flow.add{
    type = "button",
    name = "new-game-plus-import-confirm-button",
    style = "confirm_button",
    caption = {"gui.confirm"}
  }
end

return gui


map_settings_gui.lua
local util = require("__NewGamePlus__/utilities")
local MOD_PREFIX = "new-game-plus-"
local GUI_PREFIX = "map-settings-"
local ENTIRE_PREFIX = MOD_PREFIX .. GUI_PREFIX
local map_settings_gui = {}

map_settings_gui.make_pollution_settings = function(parent, map_settings)
  local WIDGET_PREFIX = "pollution-"
  local flow = parent.add{
    type = "flow",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow",
    direction = "vertical"
  }
  flow.add{
    type = "label",
    caption = {"gui-map-generator.pollution"},
    style = "caption_label"
  }
  local table = flow.add{
    type = "table",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "table",
    column_count = 2,
    style = "bordered_table"
  }
  table.style.column_alignments[2] = "center"

  table.add{
    type = "label",
    caption = {"gui-map-generator.pollution"}
  }
  table.add{
    type = "checkbox",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox",
    state = map_settings.pollution.enabled,
  }
  table.children[1].style.horizontally_stretchable = true
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "dissipation", {"gui-map-generator.pollution-absorption-modifier"}, {"gui-map-generator.pollution-absorption-modifier-description"}, tostring(map_settings.pollution.ageing), 50)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "consumption", {"gui-map-generator.enemy-attack-pollution-consumption-modifier"}, {"gui-map-generator.enemy-attack-pollution-consumption-modifier-description"}, tostring(map_settings.pollution.enemy_attack_pollution_consumption_modifier), 50)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "tree-dmg", {"gui-map-generator.minimum-pollution-to-damage-trees"}, {"gui-map-generator.minimum-pollution-to-damage-trees-description"}, tostring(map_settings.pollution.min_pollution_to_damage_trees), 50)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "tree-absorb", {"gui-map-generator.pollution-absorbed-per-tree-damaged"}, {"gui-map-generator.pollution-absorbed-per-tree-damaged-description"}, tostring(map_settings.pollution.pollution_restored_per_tree_damage), 50)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "diffusion", {"gui." .. MOD_PREFIX .. "in-unit", {"gui-map-generator.pollution-diffusion-ratio"}, {"gui." .. MOD_PREFIX .. "percent"}}, {"gui-map-generator.pollution-diffusion-ratio-description"}, tostring(map_settings.pollution.diffusion_ratio * 100), 50)
end

map_settings_gui.make_evolution_settings = function(parent, map_settings)
  local WIDGET_PREFIX = "evolution-"
  local flow = parent.add{
    type = "flow",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow",
    direction = "vertical"
  }
  flow.add{
    type = "label",
    caption = {"gui-map-generator.evolution"},
    style = "caption_label"
  }
  local table = flow.add{
    type = "table",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "table",
    column_count = 2,
    style = "bordered_table"
  }
  table.style.column_alignments[2] = "center"

  table.add{
    type = "label",
    caption = {"gui-map-generator.evolution"}
  }
  table.add{
    type = "checkbox",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox",
    state = map_settings.enemy_evolution.enabled,
  }
  table.children[1].style.horizontally_stretchable = true
  -- Warning: Below is excluded only here, not in change map settings
  -- map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "factor", {"gui-map-generator.evolution"}, {"gui." .. MOD_PREFIX .. "evolution-factor-tooltip"}, util.number_to_string(game.forces["enemy"].evolution_factor), 80)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "time", {"gui-map-generator.evolution-time-factor"}, {"gui-map-generator.evolution-time-factor-description"}, util.number_to_string(map_settings.enemy_evolution.time_factor * 100), 80)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "destroy", {"gui-map-generator.evolution-destroy-factor"}, {"gui-map-generator.evolution-destroy-factor-description"}, util.number_to_string(map_settings.enemy_evolution.destroy_factor * 100), 80)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "pollution", {"gui-map-generator.evolution-pollution-factor"}, {"gui-map-generator.evolution-pollution-factor-description"}, util.number_to_string(map_settings.enemy_evolution.pollution_factor * 100), 80)
end

map_settings_gui.make_expansion_settings = function(parent, map_settings)
  local WIDGET_PREFIX = "expansion-"
  local flow = parent.add{
    type = "flow",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow",
    direction = "vertical"
  }
  flow.add{
    type = "label",
    caption = {"gui-map-generator.enemy-expansion-group-tile"},
    style = "caption_label"
  }
  local table = flow.add{
    type = "table",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "table",
    column_count = 2,
    style = "bordered_table"
  }
  table.style.column_alignments[2] = "center"

  table.add{
    type = "label",
    caption = {"gui-map-generator.enemy-expansion-group-tile"}
  }
  table.add{
    type = "checkbox",
    name = ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox",
    state = map_settings.enemy_expansion.enabled,
  }
  table.children[1].style.horizontally_stretchable = true
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "distance", {"gui-map-generator.enemy-expansion-maximum-expansion-distance"}, {"gui-map-generator.enemy-expansion-maximum-expansion-distance-description"}, tostring(map_settings.enemy_expansion.max_expansion_distance), 30)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "min-size", {"gui-map-generator.enemy-expansion-minimum-expansion-group-size"}, {"gui-map-generator.enemy-expansion-minimum-expansion-group-size-description"}, tostring(map_settings.enemy_expansion.settler_group_min_size), 30)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "max-size", {"gui-map-generator.enemy-expansion-maximum-expansion-group-size"}, {"gui-map-generator.enemy-expansion-maximum-expansion-group-size-description"}, tostring(map_settings.enemy_expansion.settler_group_max_size), 30)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "min-cd", {"gui." .. MOD_PREFIX .. "in-unit", {"gui-map-generator.enemy-expansion-minimum-expansion-cooldown"}, {"gui." .. MOD_PREFIX .. "minutes"}}, {"gui-map-generator.enemy-expansion-minimum-expansion-cooldown-description"},  tostring(map_settings.enemy_expansion.min_expansion_cooldown / 3600), 30)
  map_settings_gui.make_config_option(table, WIDGET_PREFIX .. "max-cd", {"gui." .. MOD_PREFIX .. "in-unit", {"gui-map-generator.enemy-expansion-maximum-expansion-cooldown"}, {"gui." .. MOD_PREFIX .. "minutes"}}, {"gui-map-generator.enemy-expansion-maximum-expansion-cooldown-description"}, tostring(map_settings.enemy_expansion.max_expansion_cooldown / 3600), 30)
end

map_settings_gui.make_config_option = function(parent, name, caption, tooltip, default, max_width)
  parent.add{
    type = "label",
    caption = util.add_info_icon_to_string(caption),
    tooltip = tooltip
  }
  local child = parent.add{
    type = "textfield",
    name = ENTIRE_PREFIX .. name .. "-textfield",
    text = default,
    numeric = true,
    allow_decimal = true,
    allow_negative = false
  }
  if max_width then child.style.maximal_width = max_width end
  return child
end

map_settings_gui.expansion_reset_to_defaults = function(parent)
  local WIDGET_PREFIX = "expansion-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = true
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "distance-textfield"].text = "7"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-size-textfield"].text = "5"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-size-textfield"].text = "20"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-cd-textfield"].text = "4"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-cd-textfield"].text = "60"
end

map_settings_gui.evolution_reset_to_defaults = function(parent)
  local WIDGET_PREFIX = "evolution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = true
  -- Warning: Below is excluded only here, not in change map settings
  -- table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "factor-textfield"].text = "0"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "time-textfield"].text = "0.0004"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "destroy-textfield"].text = "0.2"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "pollution-textfield"].text = "0.000090"
end

map_settings_gui.pollution_reset_to_defaults = function(parent)
  local WIDGET_PREFIX = "pollution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = true
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "dissipation-textfield"].text = "1"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "consumption-textfield"].text = "1"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-dmg-textfield"].text = "60"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-absorb-textfield"].text = "10"
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "diffusion-textfield"].text = "2"
end

map_settings_gui.expansion_set_to_current = function(parent, map_settings)
  local WIDGET_PREFIX = "expansion-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = map_settings.enemy_expansion.enabled
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "distance-textfield"].text = tostring(map_settings.enemy_expansion.max_expansion_distance)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-size-textfield"].text = tostring(map_settings.enemy_expansion.settler_group_min_size)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-size-textfield"].text = tostring(map_settings.enemy_expansion.settler_group_max_size)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-cd-textfield"].text = tostring(map_settings.enemy_expansion.min_expansion_cooldown / 3600)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-cd-textfield"].text = tostring(map_settings.enemy_expansion.max_expansion_cooldown / 3600)
end

map_settings_gui.evolution_set_to_current = function(parent, map_settings)
  local WIDGET_PREFIX = "evolution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = map_settings.enemy_evolution.enabled
  -- Warning: Below is excluded only here, not in change map settings
  -- table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "factor-textfield"].text = util.number_to_string(game.forces["enemy"].evolution_factor)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "time-textfield"].text = util.number_to_string(map_settings.enemy_evolution.time_factor * 100)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "destroy-textfield"].text = util.number_to_string(map_settings.enemy_evolution.destroy_factor * 100)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "pollution-textfield"].text = util.number_to_string(map_settings.enemy_evolution.pollution_factor * 100)
end

map_settings_gui.pollution_set_to_current = function(parent, map_settings)
  local WIDGET_PREFIX = "pollution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state = map_settings.pollution.enabled
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "dissipation-textfield"].text = tostring(map_settings.pollution.ageing)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "consumption-textfield"].text = tostring(map_settings.pollution.enemy_attack_pollution_consumption_modifier)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-dmg-textfield"].text = tostring(map_settings.pollution.min_pollution_to_damage_trees)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-absorb-textfield"].text = tostring(map_settings.pollution.pollution_restored_per_tree_damage)
  table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "diffusion-textfield"].text = tostring(map_settings.pollution.diffusion_ratio * 100)
end

-- can throw!
map_settings_gui.expansion_read = function(parent)
  local WIDGET_PREFIX = "expansion-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  local enemy_expansion = {}

  enemy_expansion.enabled = table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state
  enemy_expansion.max_expansion_distance = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "distance-textfield"]),
                                                             2, 20,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-distance"})
  enemy_expansion.settler_group_min_size = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-size-textfield"]),
                                                             1, 20,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-min-size"})
  enemy_expansion.settler_group_max_size = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-size-textfield"]),
                                                             math.max(enemy_expansion.settler_group_min_size, 1), 50,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-max-size"})
  enemy_expansion.min_expansion_cooldown = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "min-cd-textfield"]),
                                                             1, 60,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-min-cd"}) * 3600
  enemy_expansion.max_expansion_cooldown = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "max-cd-textfield"]),
                                                             math.max(enemy_expansion.min_expansion_cooldown / 3600, 5), 180,
                                                             {"msg." .. MOD_PREFIX .. "invalid-expansion-max-cd"}) * 3600
  return enemy_expansion
end

-- can throw!
map_settings_gui.evolution_read = function(parent)
  local WIDGET_PREFIX = "evolution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  local enemy_evolution = {}

  enemy_evolution.enabled = table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state
  -- Warning: Below is excluded only here, not in change map settings
  -- enemy_evolution.evolution_factor = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "factor-textfield"]),
  --                                                           0, 1,
  --                                                           {"msg." .. MOD_PREFIX .. "invalid-evolution-factor"})
  enemy_evolution.time_factor = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "time-textfield"]),
                                                  0, 0.01,
                                                  {"msg." .. MOD_PREFIX .. "invalid-evolution-time"}) / 100
  enemy_evolution.destroy_factor = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "destroy-textfield"]),
                                                     0, 1,
                                                     {"msg." .. MOD_PREFIX .. "invalid-evolution-destroy"}) / 100
  enemy_evolution.pollution_factor = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "pollution-textfield"]),
                                                       0, 0.01,
                                                       {"msg." .. MOD_PREFIX .. "invalid-evolution-pollution"}) / 100
  return enemy_evolution
end

-- can throw!
map_settings_gui.pollution_read = function(parent)
  local WIDGET_PREFIX = "pollution-"
  local table = parent[ENTIRE_PREFIX .. WIDGET_PREFIX .. "flow"][ENTIRE_PREFIX .. WIDGET_PREFIX .. "table"]
  local pollution = {}

  pollution.enabled = table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "checkbox"].state
  pollution.ageing = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "dissipation-textfield"]),
                                       0.1, 4,
                                       {"msg." .. MOD_PREFIX .. "invalid-pollution-absorption"})
  pollution.enemy_attack_pollution_consumption_modifier = util.check_bounds(util.textfield_to_number(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "consumption-textfield"]),
                                                                            0.1, 4,
                                                                            {"msg." .. MOD_PREFIX .. "invalid-enemy-attack-pollution-consumption"})
  pollution.min_pollution_to_damage_trees = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-dmg-textfield"]),
                                                              0, 9999,
                                                              {"msg." .. MOD_PREFIX .. "invalid-pollution-tree-dmg"})
  pollution.pollution_restored_per_tree_damage = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "tree-absorb-textfield"]),
                                                                   0, 9999,
                                                                   {"msg." .. MOD_PREFIX .. "invalid-pollution-tree-absorb"})
  pollution.diffusion_ratio = util.check_bounds(util.textfield_to_uint(table[ENTIRE_PREFIX .. WIDGET_PREFIX .. "diffusion-textfield"]),
                                                0, 25,
                                                {"msg." .. MOD_PREFIX .. "invalid-pollution-diffusion"}) / 100
  return pollution
end

return map_settings_gui
